In [60]:
# Imports and pip installations (if needed)pd.reapd.pd
import pandas as pd
import numpy as np

# Part 1: Load the dataset

In [61]:

# Load the given datasets
categorical = pd.read_csv('data/chronic_kidney_disease_categorical.csv')
numerical   = pd.read_csv('data/chronic_kidney_disease_numerical.csv')
# Print the data
print(categorical)
print(numerical)

     unique_id   al   su       rbc        pc         pcc          ba  htn  \
0       224481  3.0  NaN       NaN       NaN  notpresent  notpresent  yes   
1       992643  0.0  0.0       NaN       NaN  notpresent  notpresent   no   
2       308740  1.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   
3       450314  1.0  0.0       NaN    normal  notpresent  notpresent   no   
4       881763  0.0  0.0    normal    normal  notpresent  notpresent   no   
..         ...  ...  ...       ...       ...         ...         ...  ...   
425     769512  0.0  0.0    normal    normal  notpresent  notpresent   no   
426     872034  2.0  1.0  abnormal  abnormal  notpresent     present  yes   
427     643007  NaN  NaN       NaN    normal  notpresent  notpresent  yes   
428     840392  NaN  NaN       NaN       NaN  notpresent  notpresent  yes   
429     856244  0.0  0.0       NaN    normal  notpresent  notpresent  yes   

      dm  cad appet   pe  ane  Target  
0    yes   no  good  yes   no     c

# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer:
This data contains information about different aspects of patients health, mostly including various biological data. Some of the features include age, blood pressure, and sugar levels. All of these biological data are the features used to predict the label of having Chronic Kidney Disease or not. The features of the dataset are both categorical and numerical. Some categorical columns are red blood cells, albumin, and sugar. Some numerical columns are age, blood pressure, and sodium.

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [62]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows

# Checks if a row has 
print(f'Numerical has {numerical.duplicated().sum()} duplicate rows')

# Drop these duplicate rows
numerical = numerical.drop_duplicates()

# Repeat the same for categorical dataset. Print the duplicate rows and drop them
print(f'Categorical has {categorical.duplicated().sum()} duplicate rows')
categorical = categorical.drop_duplicates()


Numerical has 30 duplicate rows
Categorical has 30 duplicate rows


## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answer:

No, it is not the case that all unique ids are present in both datasets. By printing out the number of rows in the numerical and categorical data, we can see that the categorical data has 10 more rows than the numerical. This means, the categorical data has 10 unique ids that are not present in the numerical data. Since we combine the datasets using merge, which performs what is equivalent to a SQL Inner Join, any rows in one that don't have corresponding ids in the other table are removed. So the 10 unique ids that are present in the categorical data but not in the numerical data are dropped in the combined table.

In [63]:
# Merge the two given numerical and categorical datasets based on their unique_ID.
merged_data = pd.merge(numerical, categorical, on='unique_id')
#Print the combined dataset
print(merged_data)

# Checking number of rows in each dataset to help answer the above question
print(f'Numerical has {len(numerical)} rows\nCategorical has {len(categorical)} rows\nMerged has {len(merged_data)} rows')

     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  \
0       203694  38.0   80.0   99.0   19.0  0.5  147.0  3.5  13.6  44.0   
1       938027  43.0   60.0  108.0   25.0  1.0  144.0  5.0  17.8  43.0   
2       421471  37.0   60.0  111.0   35.0  0.8  135.0  4.1  16.2  50.0   
3       764115  70.0   90.0  144.0  125.0  4.0  136.0  4.6  12.0  37.0   
4       240975  47.0   80.0   95.0   35.0  0.9  140.0  4.1   NaN   NaN   
..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...   
385     522912  56.0   80.0  165.0   55.0  1.8    NaN  NaN  13.5  40.0   
386     475200  64.0  100.0  163.0   54.0  7.2  140.0  4.6   7.9  26.0   
387     168242  48.0   60.0  112.0   44.0  1.2  142.0  4.9  14.5  44.0   
388     474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0   
389     926033  50.0   70.0  230.0   50.0  2.2    NaN  NaN  12.0  41.0   

        wbcc  rbcc   al   su       rbc        pc         pcc          ba  htn  \
0     7300.0   6.4  0.0  0.0  

## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [64]:
# Calculate the percentage of rows that contain atleast one missing value
missing = merged_data.isna().any(axis=1).sum()
percent_missing = 100 * missing / len(merged_data)
# Print %
print(f'Rows with missing data: {percent_missing}%')
# Drop these rows from the dataset
dataset = merged_data.dropna()
# Print the Dataset
print(dataset)

Rows with missing data: 60.76923076923077%
     unique_id   age     bp    bgr    bu   sc    sod  pot  hemo   pcv  \
0       203694  38.0   80.0   99.0  19.0  0.5  147.0  3.5  13.6  44.0   
1       938027  43.0   60.0  108.0  25.0  1.0  144.0  5.0  17.8  43.0   
2       421471  37.0   60.0  111.0  35.0  0.8  135.0  4.1  16.2  50.0   
7       109053  64.0   60.0  106.0  27.0  0.7  150.0  3.3  14.4  42.0   
13      349892  59.0   70.0  424.0  55.0  1.7  138.0  4.5  12.6  37.0   
..         ...   ...    ...    ...   ...  ...    ...  ...   ...   ...   
379     486397  20.0   70.0  123.0  44.0  1.0  135.0  3.8  14.6  44.0   
381     462892  36.0   80.0   85.0  16.0  1.1  142.0  4.1  15.6  44.0   
386     475200  64.0  100.0  163.0  54.0  7.2  140.0  4.6   7.9  26.0   
387     168242  48.0   60.0  112.0  44.0  1.2  142.0  4.9  14.5  44.0   
388     474407  21.0   90.0  107.0  40.0  1.7  125.0  3.5   8.3  23.0   

        wbcc  rbcc   al   su       rbc        pc         pcc          ba  htn  \

## Part 3.4: Sort the dataset according to the Labels

In [65]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
sorted_data = dataset.sort_values(by='Target').reset_index(drop=True)
# Print the dataset
print(sorted_data)

     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  \
0       474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0   
1       137148  73.0  100.0  295.0   90.0  5.6  140.0  2.9   9.2  30.0   
2       484175  53.0   90.0   70.0  107.0  7.2  114.0  3.7   9.5  29.0   
3       343710  61.0   80.0  173.0  148.0  3.9  135.0  5.2   7.7  24.0   
4       532520  60.0   90.0  105.0   53.0  2.3  136.0  5.2  11.1  33.0   
..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...   
148     647552  61.0   70.0  120.0   29.0  0.7  137.0  3.5  17.4  52.0   
149     125587  37.0   60.0  109.0   47.0  1.1  141.0  4.9  15.0  48.0   
150     614376  57.0   80.0  119.0   17.0  1.2  135.0  4.7  15.4  42.0   
151     853426  41.0   70.0  125.0   38.0  0.6  140.0  5.0  16.8  41.0   
152     572140  55.0   80.0  104.0   28.0  0.9  142.0  4.8  17.3  52.0   

        wbcc  rbcc   al   su       rbc        pc         pcc          ba  htn  \
0    12400.0   3.9  4.0  0.0  

## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [66]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']
data_with_dummies = pd.get_dummies(sorted_data, columns=categorical_columns, dtype=int)
# Print the dataset
print(data_with_dummies)

     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  \
0       474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0   
1       137148  73.0  100.0  295.0   90.0  5.6  140.0  2.9   9.2  30.0   
2       484175  53.0   90.0   70.0  107.0  7.2  114.0  3.7   9.5  29.0   
3       343710  61.0   80.0  173.0  148.0  3.9  135.0  5.2   7.7  24.0   
4       532520  60.0   90.0  105.0   53.0  2.3  136.0  5.2  11.1  33.0   
..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...   
148     647552  61.0   70.0  120.0   29.0  0.7  137.0  3.5  17.4  52.0   
149     125587  37.0   60.0  109.0   47.0  1.1  141.0  4.9  15.0  48.0   
150     614376  57.0   80.0  119.0   17.0  1.2  135.0  4.7  15.4  42.0   
151     853426  41.0   70.0  125.0   38.0  0.6  140.0  5.0  16.8  41.0   
152     572140  55.0   80.0  104.0   28.0  0.9  142.0  4.8  17.3  52.0   

        wbcc  rbcc   al   su  rbc_abnormal  rbc_normal  pc_abnormal  \
0    12400.0   3.9  4.0  0.0            

In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

Answer: For this dataset, it would create an equivalent model if we used a single column for the binary variable. This is because all of the categorical data are binary, meaning they can only take on two values. However, if the categorical variable could take on more than 2 values, this is not the case. Saying 0 or 1 will not tell use which of the more than 2 values we might be referring to and will not let us predict values beyond 1. So in the sandwich example, we will not be able to accurately predict the type of condiment with just a single binary column as we cannot represent all 5 condiments using a single binary value.

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [67]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']
filtered_data = data_with_dummies
for col in numerical_columns:
    mean = filtered_data[col].mean()
    std = filtered_data[col].std()
    filtered_data = filtered_data[(np.abs(filtered_data[col] - mean) <= (3 * std))]
# Print the dataset
print(filtered_data[numerical_columns].head(100))

      age     bp    bgr     bu   sc    sod  pot  hemo   pcv     wbcc  rbcc
0    21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0  12400.0   3.9
1    73.0  100.0  295.0   90.0  5.6  140.0  2.9   9.2  30.0   7000.0   3.2
3    61.0   80.0  173.0  148.0  3.9  135.0  5.2   7.7  24.0   9200.0   3.2
4    60.0   90.0  105.0   53.0  2.3  136.0  5.2  11.1  33.0  10500.0   4.1
6    46.0   60.0  163.0   92.0  3.3  141.0  4.0   9.8  28.0  14600.0   3.2
..    ...    ...    ...    ...  ...    ...  ...   ...   ...      ...   ...
115  37.0   60.0  111.0   35.0  0.8  135.0  4.1  16.2  50.0   5500.0   5.7
116  43.0   60.0  108.0   25.0  1.0  144.0  5.0  17.8  43.0   7200.0   5.5
117  22.0   60.0   97.0   18.0  1.2  138.0  4.3  13.5  42.0   7900.0   6.4
118  58.0   80.0  100.0   50.0  1.2  140.0  3.5  14.0  50.0   6700.0   6.5
119  39.0   60.0   86.0   37.0  0.6  150.0  5.0  13.6  51.0   5800.0   4.5

[100 rows x 11 columns]


## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [68]:
# Normalize the all Numerical Attributes in the dataset.

# We normalize the data using Min-Max scaling
normalized_data = filtered_data
for col in numerical_columns:
    max_val = normalized_data[col].max()
    min_val = normalized_data[col].min()
    normalized_data[col] = (normalized_data[col] - min_val) / (max_val - min_val)
# Print the dataset
print(normalized_data)

     unique_id       age    bp       bgr        bu        sc       sod  \
0       474407  0.202703  0.75  0.158798  0.196078  0.160494  0.166667   
1       137148  0.905405  1.00  0.965665  0.522876  0.641975  0.666667   
3       343710  0.743243  0.50  0.442060  0.901961  0.432099  0.500000   
4       532520  0.729730  0.75  0.150215  0.281046  0.234568  0.533333   
6       546225  0.540541  0.00  0.399142  0.535948  0.358025  0.700000   
..         ...       ...   ...       ...       ...       ...       ...   
148     647552  0.743243  0.25  0.214592  0.124183  0.037037  0.566667   
149     125587  0.418919  0.00  0.167382  0.241830  0.086420  0.700000   
150     614376  0.689189  0.50  0.210300  0.045752  0.098765  0.500000   
151     853426  0.472973  0.25  0.236052  0.183007  0.024691  0.666667   
152     572140  0.662162  0.50  0.145923  0.117647  0.061728  0.733333   

          pot      hemo       pcv      wbcc      rbcc   al   su  rbc_abnormal  \
0    0.206897  0.059406  0.000

## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer: The one column that has nothing to do with the anything health related for a patient is their unique_id. This is simply just a number to uniquely identify each person's data. Including such data in our modeling and predicitions might cause them to be off as we might find meaningless correlations between id and other data.

In [69]:
#Remove that column
clean_data = normalized_data.drop(columns='unique_id')
# Print the dataset
print(clean_data)

          age    bp       bgr        bu        sc       sod       pot  \
0    0.202703  0.75  0.158798  0.196078  0.160494  0.166667  0.206897   
1    0.905405  1.00  0.965665  0.522876  0.641975  0.666667  0.000000   
3    0.743243  0.50  0.442060  0.901961  0.432099  0.500000  0.793103   
4    0.729730  0.75  0.150215  0.281046  0.234568  0.533333  0.793103   
6    0.540541  0.00  0.399142  0.535948  0.358025  0.700000  0.379310   
..        ...   ...       ...       ...       ...       ...       ...   
148  0.743243  0.25  0.214592  0.124183  0.037037  0.566667  0.206897   
149  0.418919  0.00  0.167382  0.241830  0.086420  0.700000  0.689655   
150  0.689189  0.50  0.210300  0.045752  0.098765  0.500000  0.620690   
151  0.472973  0.25  0.236052  0.183007  0.024691  0.666667  0.724138   
152  0.662162  0.50  0.145923  0.117647  0.061728  0.733333  0.655172   

         hemo       pcv      wbcc      rbcc   al   su  rbc_abnormal  \
0    0.059406  0.000000  0.653226  0.257143  4.0  0.

## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [70]:
# Export the dataframe to a new csv file
clean_data.to_csv('chronic_kidney_disease.csv')
# Export the dataframe to a new json file
clean_data.to_json('chronic_kidney_disease.json')

# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

In [71]:
# Printing top 15 rows
pd.set_option('display.max_columns', None)
print(clean_data.head(15))

         age    bp       bgr        bu        sc       sod       pot  \
0   0.202703  0.75  0.158798  0.196078  0.160494  0.166667  0.206897   
1   0.905405  1.00  0.965665  0.522876  0.641975  0.666667  0.000000   
3   0.743243  0.50  0.442060  0.901961  0.432099  0.500000  0.793103   
4   0.729730  0.75  0.150215  0.281046  0.234568  0.533333  0.793103   
6   0.540541  0.00  0.399142  0.535948  0.358025  0.700000  0.379310   
8   0.675676  0.75  0.253219  0.633987  0.777778  0.366667  0.655172   
10  0.716216  0.50  1.000000  0.163399  0.111111  0.066667  0.206897   
13  0.729730  0.00  0.935622  0.169935  0.160494  0.333333  0.034483   
17  0.000000  0.00  0.103004  0.372549  0.074074  0.500000  0.689655   
19  0.878378  0.00  0.206009  0.751634  0.604938  0.533333  0.689655   
20  0.851351  0.25  0.618026  0.562092  0.728395  0.000000  0.344828   
21  0.878378  0.25  0.639485  0.470588  0.395062  0.433333  0.517241   
22  0.783784  0.00  0.725322  0.313725  0.481481  0.566667  0.86

### Paste the markdown table here

| | age | bp | bgr | bu | sc | sod | pot | hemo | pcv | wbcc | rbcc | al | su | rbc_abnormal | rbc_normal | pc_abnormal | pc_normal | pcc_notpresent | pcc_present | ba_notpresent | ba_present | htn_no | htn_yes | dm_no | dm_yes | cad_no | cad_yes | appet_good | appet_poor | pe_no | pe_yes | ane_no | ane_yes | Target_ckd | Target_notckd |
|----|----------|-------|-----------|-----------|-----------|-----------|-----------|------------|-----------|-----------|-----------|------|------|----------------|--------------|---------------|-------------|------------------|---------------|-----------------|--------------|----------|-----------|---------|----------|----------|-----------|---------------|---------------|---------|----------|----------|-----------|---------------|-----------------|
| 0 | 0.202703 | 0.75 | 0.158798 | 0.196078 | 0.160494 | 0.166667 | 0.206897 | 0.059406 | 0.000000 | 0.653226 | 0.257143 | 4.0 | 0.0 | 0 | 1 | 1 | 0 | 0 | 1 | 0 | 1 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 0 | 1 | 1 | 0 |
| 1 | 0.905405 | 1.00 | 0.965665 | 0.522876 | 0.641975 | 0.666667 | 0.000000 | 0.148515 | 0.225806 | 0.217742 | 0.057143 | 3.0 | 2.0 | 1 | 0 | 1 | 0 | 0 | 1 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 1 | 0 | 1 | 1 | 0 | 1 | 0 | 1 | 0 |
| 3 | 0.743243 | 0.50 | 0.442060 | 0.901961 | 0.432099 | 0.500000 | 0.793103 | 0.000000 | 0.032258 | 0.395161 | 0.057143 | 2.0 | 0.0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 1 | 0 | 1 | 0 | 1 | 1 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 1 | 0 |
| 4 | 0.729730 | 0.75 | 0.150215 | 0.281046 | 0.234568 | 0.533333 | 0.793103 | 0.336634 | 0.322581 | 0.500000 | 0.314286 | 2.0 | 0.0 | 1 | 0 | 0 | 1 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 1 | 0 | 1 | 1 | 0 |
| 6 | 0.540541 | 0.00 | 0.399142 | 0.535948 | 0.358025 | 0.700000 | 0.379310 | 0.207921 | 0.161290 | 0.830645 | 0.057143 | 1.0 | 0.0 | 0 | 1 | 0 | 1 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 1 | 0 | 1 | 1 | 0 |
| 8 | 0.675676 | 0.75 | 0.253219 | 0.633987 | 0.777778 | 0.366667 | 0.655172 | 0.138614 | 0.193548 | 0.169355 | 0.114286 | 2.0 | 0.0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 1 | 0 | 1 | 1 | 0 |
| 10 | 0.716216 | 0.50 | 1.000000 | 0.163399 | 0.111111 | 0.066667 | 0.206897 | 0.267327 | 0.387097 | 0.532258 | 0.371429 | 1.0 | 0.0 | 1 | 0 | 0 | 1 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1 | 0 | 0 | 1 | 1 | 1 | 0 | 1 | 1 | 0 |
| 13 | 0.729730 | 0.00 | 

** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

Unfortuantely after prompting 383GPT several times, it begins to hallucinate halfway through generating the markdown table. It seems as though the more data provided to the LLM will make it more likely to hallucinate. So to mitigate data conversion hallucinations, we can limit the amount of data that we give the LLM. For example, we could give it 5 rows of data instead of 15.

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [72]:
# Code to rename all the columns in the dataset
# Mapping of abbreviations to full names
abbr_to_full = {
    'age': 'age',
    'bp': 'blood pressure',
    'sg': 'specific gravity',
    'al': 'albumin',
    'su': 'sugar',
    'rbc': 'red blood cells',
    'pc': 'pus cell',
    'pcc': 'pus cell clumps',
    'ba': 'bacteria',
    'bgr': 'blood glucose random',
    'bu': 'blood urea',
    'sc': 'serum creatinine',
    'sod': 'sodium',
    'pot': 'potassium',
    'hemo': 'hemoglobin',
    'pcv': 'packed cell volume',
    'wc': 'white blood cell count',
    'rc': 'red blood cell count',
    'htn': 'hypertension',
    'dm': 'diabetes mellitus',
    'cad': 'coronary artery disease',
    'appet': 'appetite',
    'pe': 'pedal edema',
    'ane': 'anemia',
    'class': 'class'
}

# Rename the columns
clean_data.rename(columns=abbr_to_full, inplace=True)

# Display the updated DataFrame
print(clean_data)

          age  blood pressure  blood glucose random  blood urea  \
0    0.202703            0.75              0.158798    0.196078   
1    0.905405            1.00              0.965665    0.522876   
3    0.743243            0.50              0.442060    0.901961   
4    0.729730            0.75              0.150215    0.281046   
6    0.540541            0.00              0.399142    0.535948   
..        ...             ...                   ...         ...   
148  0.743243            0.25              0.214592    0.124183   
149  0.418919            0.00              0.167382    0.241830   
150  0.689189            0.50              0.210300    0.045752   
151  0.472973            0.25              0.236052    0.183007   
152  0.662162            0.50              0.145923    0.117647   

     serum creatinine    sodium  potassium  hemoglobin  packed cell volume  \
0            0.160494  0.166667   0.206897    0.059406            0.000000   
1            0.641975  0.666667   0.000

## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [73]:
# Converted SQL to Pandas code

# Changed 'Target' to 'Target_ckd' to avoid keyError, but the rest comes from 383GPT
result = clean_data.groupby('Target_ckd').size().reset_index(name='count')
print(result)

   Target_ckd  count
0           0    111
1           1     22


*Describe what the above code does here*

The query above simply will group together rows that have the same value for the column 'Target_ckd' and diplay the count for the number of rows that have that value in the 'Target_ckd' column. The output here tells us 111 rows have 0 in the column 'Target_ckd' while 22 have 1, meaning 22 people have Chronic Kidney Disease in this dataset.